### ISSUE TITLE
Framework to add jittering to the data on the level of the _machine intrinsic_ bin sizes. 

### ISSUE THREAD

**Issue by steinnymir** [OpenCOMPES/sed#6]
This requires passing different jitter sizes for different dimensions.

See MPES

**Comment by steinnymir** [comment]
Together with #7, this requires to define jittering methods.
One option is to add an automatic application of jitter to digitized axes, when provided with the minimum digital step.
An other path which seems to improve performance in MPES is to re-use the same arrays of randoms for all axes, which does not require to re-create them each time.

**Comment by rettigl** [comment]
As I commented in the pull request #50, the difficulty lies in knowing the original machine bin size for derived data frame colums, in particular when there are nonlinear/complex operations involved, like for the distortion correction / image registration / momentum calibration

**Comment by steinnymir** [comment]
You are right, automatic evaluation can be tricky if done after processing.
One approach would be to apply the jitter before any processing, which is what we were doing in hextof-processor.
I would try implment that, and eventually a method that can read through the data and guess the best step, even though it might be computationally expensive. This function could be used when the step is not provided in the original metadata.

**Comment by rettigl** [comment]
If you can do this without actually reading the data, great. I looked into this once, and failed. you don't know the size of the data frame partition before actually loading the data...

**Comment by steinnymir** [comment]
Without reading the data is impossible, but can make an educated guess based on a single partition, for example.
This also needs to be done only once at the data loading step.

Even better though would be to save such information as metadata for each column when generating the dataframe. There we are anyway reading the data, and gathering some statistical information could be very valuable, such as mean, max, min, minimum step, number of values etc...
This however would not be done in this package but in the experiment specific ones (hextof-loader, MPES-loader...)

**Comment by zainsohail04** [comment]
> Without reading the data is impossible, but can make an educated guess based on a single partition, for example. This also needs to be done only once at the data loading step.
> 
> Even better though would be to save such information as metadata for each column when generating the dataframe. There we are anyway reading the data, and gathering some statistical information could be very valuable, such as mean, max, min, minimum step, number of values etc... This however would not be done in this package but in the experiment specific ones (hextof-loader, MPES-loader...)

Coming from the PR and after reading this, I think it'd be quite useful if we can write down a set of suggestions for the experiment specific packages, on what they can include in the metadata for the best possible integration. Since I am working on hextofloader, naturally, I can keep those things in mind and we can reiterate such suggestions and perfect it. What do you think?

**Comment by steinnymir** [comment]
I created a discussion #59 in which we can collect all suggestions for the structure and metadata to include in the dataframes.

**Comment by rettigl** [comment]
After the discussion today, and a quick test, to me it seems the best option to really implement the jittering on the level of the initial machine output data, before performing any conversion in the workflow, with map_partitions. This indeed works nicely in a delayed fashion. The only downside I see is that you need to decide for or against jittering at the beginning of defining the workflow (or duplicating the whole workflow once jittered, once non-jittered, blowing up the dataframe), and that binning without jitter then still has the computational overhead of jittering. Also, I could not get simultaneous jittering on multiple axes within one run working yet

**Comment by rettigl** [comment]
This function allows to apply the jittering to multiple columns simultaneously. According to my tests, the speedup is almost a factor of 2 for a 3-dimensional binning. We need to decide where to put this, and if we want to keep the jittering option in the binning function. Do we make a separate file for "dataframe operations", or does this go into calibration.py?

```
import numpy as np
def apply_jitter(df, amps, cols, cols_jittered, type):
    """ Add jittering to a dataframe column.
    :Parameters:
        df : dataframe
            Dataframe to add noise/jittering to.
        amps : list 
            Amplitude scalings for the jittering noise.
        cols : str list
            Names of the columns to add jittering to.
        cols_jittered : str list
            Names of the columns with added jitter.
    :Return:
        dataframe with added columns
    """
    colsize = df[cols[0]].size

    jitter = np.zeros(colsize)

    if (type == 'uniform'):
        # Uniform Jitter distribution
        jitter =  np.random.uniform(low=-1, high=1, size=colsize)
    elif (type == 'normal'):
        # Normal Jitter distribution works better for non-linear transformations and jitter sizes that don't match the original bin sizes
        jitter =  np.random.standard_normal(size=colsize)

    for (col, col_jittered, amp) in zip(cols, cols_jittered, amps):
        df[col_jittered] = df[col] + amp * jitter

    return df
```

**Comment by rettigl** [comment]
As one would expect, applying uniform jitter with 0.5* the intrinsic machine bin size completely removes any digitization artefacts, even for binning grids finer than the machine grid. This is what we should set as the standard approach.

**Comment by steinnymir** [comment]
Great to hear the jitter on multiple columns works!

> Do we make a separate file for "dataframe operations", or does this go into calibration.py?

I would suggest to make a dfops.py module in sed/sed where to put this kind of functions, as I don't think this will be the last.
Before anyone complains, make sure to format the function according to our coding standards. 😄 


> This is what we should set as the standard approach.

We could try set it automatic for columns of digitized values, based on provided metadata. We need to think how to make this work, but it would be really nice to have. 
It could be usefull to also append more information on the columns in a similar way, such as descriptions of what they represent, for example.

**Comment by rettigl** [comment]
I implemented an approach to this in PR #69
